src/core/thread/dua_manager.cpp

```diff
@@ -66,33 +68,110 @@ void DuaManager::UpdateDomainUnicastAddress(BackboneRouter::Leader::DomainPrefix
-    mDomainUnicastAddress.GetAddress().SetPrefix(prefix->mPrefix.mFields.m8, prefix->mLength);
     mDomainUnicastAddress.mPrefixLength = prefix->mLength;
 
-    if (Get<Utils::Slaac>().GenerateIid(mDomainUnicastAddress, NULL, 0, &mDadCounter,
-                                        false /* not allow random IID if no valid IID */) == OT_ERROR_NONE)
+    // Apply cached DUA manually configured when prefixes match.
+    if (!mFixedDomainUnicastAddress.IsUnspecified() &&
+        Ip6::Address::PrefixMatch(mFixedDomainUnicastAddress.mFields.m8, prefix->mPrefix.mFields.m8, prefix->mLength) >=
+            prefix->mLength)
     {
-        mDomainUnicastAddress.mValid = true;
+        mDomainUnicastAddress.GetAddress() = mFixedDomainUnicastAddress;
+        mDomainUnicastAddress.mValid       = true;
         Get<ThreadNetif>().AddUnicastAddress(mDomainUnicastAddress);
     }
     else
+    {
+        mDomainUnicastAddress.GetAddress().Clear();
+        mDomainUnicastAddress.GetAddress().SetPrefix(prefix->mPrefix.mFields.m8, prefix->mLength);
+
+        if (GenerateDomainUnicastAddressIid() == OT_ERROR_NONE)
+        {
+            Get<ThreadNetif>().AddUnicastAddress(mDomainUnicastAddress);
+        }
+    }
+
+exit:
+    return;
+}
+
+otError DuaManager::GenerateDomainUnicastAddressIid(void)
+{
+    otError error;
+
+    if ((error = Get<Utils::Slaac>().GenerateIid(mDomainUnicastAddress, NULL, 0, &mDadCounter,
+                                                 false /* not allow random IID if no valid IID */)) == OT_ERROR_NONE)
+    {
+        mDomainUnicastAddress.mValid = true;
+        otLogInfoCore("Generated DUA: %s", mDomainUnicastAddress.GetAddress().ToString().AsCString());
+    }
+    else
     {
         otLogWarnCore("Failed to generate valid DUA");
     }
 
+    return error;
+}
+
+void DuaManager::SetFixedDomainUnicastAddress(const Ip6::Address &aAddress)
+{
+    mFixedDomainUnicastAddress = aAddress;
+
+    // No action if the manually specified address is the same as what is added.
+    VerifyOrExit(aAddress != mDomainUnicastAddress.GetAddress(), OT_NOOP);
+
+    if (Get<Mle::MleRouter>().IsAttached() && Get<ThreadNetif>().IsUnicastAddress(mDomainUnicastAddress.GetAddress()))
+    {
+        // Skip if the specified address has a different domain prefix.
+        VerifyOrExit(mFixedDomainUnicastAddress.PrefixMatch(mDomainUnicastAddress.GetAddress()) >=
+                         mDomainUnicastAddress.mPrefixLength,
+                     OT_NOOP);
+
+        // Update the Domain Unicast Address.
+        Get<ThreadNetif>().RemoveUnicastAddress(mDomainUnicastAddress);
+        mDomainUnicastAddress.GetAddress() = mFixedDomainUnicastAddress;
+        Get<ThreadNetif>().AddUnicastAddress(mDomainUnicastAddress);
+    }
+
 exit:
     return;
 }
 
+void DuaManager::ClearFixedDomainUnicastAddress(void)
+{
+    if (!mFixedDomainUnicastAddress.IsUnspecified() && Get<ThreadNetif>().IsUnicastAddress(mFixedDomainUnicastAddress))
```

Comment:
Just checking if the fixed DUA is a unicast address does not make sure that it is also the active dua?
Following commands can cause unnecessary removal and regeneration of DUA ?

```bash
node 1> ipaddr
fdde:ad00:beef:0:0:0:0:fc38
fdde:ad00:beef:0:0:ff:fe00:fc10
fd00:7d03:7d03:7d03:15e5:ca8:f5b:d313
fdde:ad00:beef:0:0:ff:fe00:fc00
fdde:ad00:beef:0:0:ff:fe00:a400
fdde:ad00:beef:0:fe30:468d:bee3:24b7
fe80:0:0:0:748d:41fd:1042:ee33
Done
node 1> dua addr fdde:ad00:beef:0:fe30:468d:bee3:24b7  # set the MLEID to be fixed DUA
Done
node 1> dua clear # then clear it
Done

```

Reply:
Good catch. will enhance.